### PR TITLE
[Scheduler] Priority scheduling supports preemption of requests in the running queue by requests in the waiting queue

### LINKED ISSUE #40004: [Feature]: Priority scheduling supports preemption of requests in the running queue by requests in the waiting queue
### 🚀 The feature, motivation and pitch

The current priority scheduling only supports evicting low-priority requests from the running queue when resources are insufficient. However, when scheduling the waiting queue, if pending requests cannot be scheduled (for example, when the number of requests in the running queue has reached max_num_seqs), even high-priority requests cannot preempt requests in the running queue.

### Alternatives

_No response_

### Additional context

_No response_

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by injaeryou**
## Purpose
Priority scheduling supports preemption of requests in the running queue by requests in the waiting queue

Fixes: #40004

## Test Plan

```bash
pytest tests/v1/core/test_scheduler.py::test_priority_preemption_at_max_num_seqs
```

## Test Result

```bash
======================================================== test session starts =========================================================
platform linux -- Python 3.12.13, pytest-9.0.3, pluggy-1.6.0
rootdir: /home/name/.test/.gpu/vllm
configfile: pyproject.toml
plugins: anyio-4.13.0
collected 1 item                                                                                                                     

tests/v1/core/test_scheduler.py .                                                                                              [100%]

========================================================== warnings summary ==========================================================
<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute

<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyObject has no __module__ attribute

.venv/lib/python3.12/site-packages/torch/jit/_script.py:365: 14 warnings
  /home/name/.test/.gpu/vllm/.venv/lib/python3.12/site-packages/torch/jit/_script.py:365: DeprecationWarning: `torch.jit.script_method` is deprecated. Please switch to `torch.compile` or `torch.export`.
    warnings.warn(

tests/v1/core/test_scheduler.py::test_priority_preemption_at_max_num_seqs
  /home/name/.test/.gpu/vllm/.venv/lib/python3.12/site-packages/transformers/models/gpt2/tokenization_gpt2.py:110: DeprecationWarning: Deprecated in 0.9.0: BPE.__init__ will not create from files anymore, try `BPE.from_file` instead
    BPE(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=================================================== 1 passed, 17 warnings in 3.86s ===================================================
sys:1: DeprecationWarning: builtin type swigvarlink has no __module__ attribute
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by sts07142** [comment]
@robertgshaw2-redhat Please review it.

**Comment by kelliaao** [comment]
This PR only adds preemption for the waiting queue when the max_num_seqs limit is reached. Could preemption for requests when other resource limits are reached be added, such as for the KV cache?

**Comment by kelliaao** [comment]
This PR only adds preemption for the waiting queue when the max_num_seqs limit is reached. Could preemption for requests when other resource limits are reached be added, such as for the KV cache?

**Comment by kelliaao** [comment]
This PR only adds preemption for the waiting queue when the max_num_seqs limit is reached. Could preemption for requests when other resource limits are reached be added, such as for the KV cache?

**Comment by kelliaao** [comment]
This PR only adds preemption for the waiting queue when the max_num_seqs limit is reached. Could preemption for requests when other resource limits are reached be added, such as for the KV cache?

**Comment by kelliaao** [comment]
This PR only adds preemption for the waiting queue when the max_num_seqs limit is reached. Could preemption for requests when other resource limits are reached be added, such as for the KV cache?

**Comment by injaeryou** [comment]
@kelliaao extended to cover KV-cache limits too.
